### PR TITLE
ipsec: Add configurable jitter to key file watcher to prevent thundering herd

### LINKED ISSUE #42721: CFP: Configurable Jitter for IPsec Key Rotation
### Is your proposed feature related to a problem?

Yes. In large production clusters (e.g., with hundreds of nodes) using Cilium with IPsec encryption enabled, weekly or periodic IPsec key rotations trigger a "thundering herd" effect. All Cilium agents detect the key file change (/etc/ipsec/keys) simultaneously via the key file watcher. The process looks like: 
- The agent's [keyfilewatcher](https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec/ipsec_linux.go#L1285) method detects changes to `/etc/ipsec/keys`
- New keys are loaded by [loadIPSecKeysFile](https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec/ipsec_linux.go#L1061)
- [SetIPsecKeyIdentity()](https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec/ipsec_linux.go#L1243) is then called.
- [SetIPsecKeyIdentity()](https://github.com/cilium/cilium/blob/main/pkg/node/address.go#L478) gets the latest key id
- Then the agent calls NodeHandler's [AllNodeValidateImplementation](https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/node.go#L767) method
- Which triggers a [nodeUpdate](https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/node.go#L777) for the local node, which sends events to nodeDiscovery
- [updateCiliumNodeResource()](https://github.com/cilium/cilium/blob/main/pkg/nodediscovery/nodediscovery.go#L222) method updates the CiliumNode resource and logs:

  ```
    n.logger.Info(
		"Creating or updating CiliumNode resource",
		logfields.Node, nodeTypes.GetName(),
    ```

- This update to the CiliumNode resource is consumed by the K8sCiliumNodeWatcher and its [onCiliumNodeUpdate()](https://github.com/cilium/cilium/blob/main/pkg/k8s/watchers/cilium_node.go#L132) method is called, which in turn calls onCiliumNodeInsert, which finally calls the [NodeUpdated()](https://github.com/cilium/cilium/blob/main/pkg/k8s/watchers/cilium_node.go#L128) method for each node. 
- This is the method that logs the "Node updated" event for each node.

  ```
    func (m *manager) NodeUpdated(n nodeTypes.Node) {
	m.logger.Info(
		"Node updated",
		logfields.ClusterName, n.Cluster,
		logfields.NodeName, n.Name,
		logfields.SPI, n.EncryptionKey,
    )
    ```

 This results in a massive spike of "Node updated" events (amplified by the number of control plane Cilium pods), overwhelming the Kubernetes API server and causing significant CPU spikes on control plane nodes. The issue is exacerbated in scaled deployments where the number of nodes amplifies the event volume, potentially leading to degraded control plane during rotations if the nodes are not vertically scaled appropriately. While key rotations are necessary for security, the synchronous nature of the watcher makes it challenging to manage in large-scale production environments.

### Describe the feature you'd like?
Introduce a configurable jitter in the IPsec key file watcher to stagger key reloading across Cilium agents, reducing the peak load on the API server during rotations. The jitter should introduce a random delay (between 0 and a configurable maximum) after detecting a key file change, before proceeding with key loading and CiliumNode updates. This would spread out the "Node Updated" events over time, mitigating CPU spikes and API overload in large clusters. Key requirements:
	• The maximum jitter duration should be configurable via Helm values or Cilium config (e.g., encryption.ipsec.jitterMaxDuration), with a sensible default (e.g., 50% of encryption.ipsec.keyRotationDuration to align with existing rotation windows).
	• Jitter should be per-agent (randomized independently) to ensure even distribution without requiring cluster-wide coordination.
	• Preserve existing behavior for small clusters or when jitter is disabled (e.g., via a flag like encryption.ipsec.enableJitter: false).
	• Ensure the jitter does not interfere with security guarantees, such as timely adoption of new keys within the keyRotationDuration window.

### Brief description of proposed solution
The implementation could modify the keyfileWatcher function in `pkg/datapath/linux/ipsec/ipsec_linux.go` to insert a jitter-based sleep after detecting a file change (Create or Write event) but before loading the keys via `loadIPSecKeysFile`. This keeps the change localized to the watcher loop without altering the core key loading or node update logic.
Example pseudocode for the addition:
```
func (a *Agent) keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath string, nodeHandler types.NodeHandler, health cell.Health) error {
    for {
        select {
        case event := <-watcher.Events:
            if event.Op&(fswatcher.Create|fswatcher.Write) == 0 {
                continue
            }
            // Existing pre-jitter logic (e.g., validation)...
//Introduce jitter: random duration between 0 and maxJitter
            maxJitter := a.GetIPSecJitterMaxDuration() // Fetched from config, e.g., 50% of keyRotationDuration
            jitter := time.Duration(rand.Int63n(int64(maxJitter)))
            time.Sleep(jitter)
//Existing post-jitter logic
            _, spi, err := a.loadIPSecKeysFile(keyfilePath)
            if err != nil {
                // Handle error...
            }
            // Proceed with SetIPsecKeyIdentity, node updates, etc.
        }
        // Rest of loop...
    }
}
```
	• Configuration: Add jitterMaxDuration to the IPsec config struct, defaulting to keyRotationDuration / 2. Expose it in Helm charts under encryption.ipsec.
	• Rationale for jitter placement: Delaying after detection but before loading ensures agents still pick up the new key promptly, while spreading the downstream CiliumNode PATCH requests and "Node Updated" logs.
	• Edge cases:
		○ If jitter exceeds keyRotationDuration, cap it or log a warning.
		○ Test in large-scale simulations to verify even distribution.
		○ Compatibility: Backwards-compatible with existing Cilium versions; no impact on non-IPsec setups.
	• Alternatives considered: Disabling the key watcher requires manual restarts, which is disruptive in production. Spreading rotations at the secret level isn't feasible for shared symmetric keys.
This approach draws from [existing jitter usage](https://github.com/cilium/cilium/pull/42282) in Cilium to avoid introducing novel patterns. I'd be happy to prototype this or discuss further.

Notify relevant community channels

Notify the members of any relevant code owners below from the [teams] list in the following form:

@cilium/sig-k8s
@cilium/ipsec

### PR DESCRIPTION
**Pull request by shavyshetty**
Adds a configurable jitter delay to the IPsec key file watcher to prevent a thundering herd on the Kubernetes API server during key rotation in large clusters. 

```release-note
Add jitter delay to the IPsec key file watcher for key rotations, to avoid thundering herd problem on Cilium agents.
```
Fixes: #42721

### CODE REVIEW

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 1]
```diff
@@ -745,6 +745,9 @@ data:
     {{- if .Values.encryption.ipsec.keyRotationDuration }}
   ipsec-key-rotation-duration: {{ include "validateDuration" .Values.encryption.ipsec.keyRotationDuration | quote }}
     {{- end }}
+    {{- if .Values.encryption.ipsec.keyRotationJitter}}
+    ipsec-key-rotation-jitter: {{ include "validateDuration" .Values.encryption.ipsec.keyRotationJitter | quote}}
```
Comment (adkafka) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: Indent looks off here, no?
```suggestion
  ipsec-key-rotation-jitter: {{ include "validateDuration" .Values.encryption.ipsec.keyRotationJitter | quote}}
```

File: pkg/datapath/linux/ipsec/cell.go [hunk 2]
```diff
@@ -104,6 +106,7 @@ func (def UserConfig) Flags(flags *pflag.FlagSet) {
 	flags.MarkHidden(types.DNSProxyInsecureSkipTransparentModeCheck)
 	flags.String(types.IPSecKeyFile, def.IPsecKeyFile, "Path to IPsec key file")
 	flags.Duration(types.IPsecKeyRotationDuration, def.IPsecKeyRotationDuration, "Maximum duration of the IPsec key rotation. The previous key will be removed after that delay.")
+	flags.Duration(types.IPsecKeyRotationJitter, def.IPsecKeyRotationJitter, "Maxium jitter duration after detecting key file change before loading keys. Prevents thundering herd on K8s API during rotations in large clusters. Disabled by default.")
```
Comment (adkafka) [inline pkg/datapath/linux/ipsec/cell.go]: Is it disabled by default, or is it set to 0 (which will set it to 50% of key rotation duration)?

File: pkg/datapath/linux/ipsec/cell.go [hunk 3]
```diff
@@ -120,6 +123,13 @@ func (c Config) DNSProxyInsecureSkipTransparentModeCheckEnabled() bool {
 	return c.DNSProxyInsecureSkipTransparentModeCheck
 }
 
+// EffectiveKeyRotationJitter returns the jitter duration to apply after detecting
+// a key file change before loading the new keys. If the configured jitter is 0,
```
Comment (adkafka) [inline pkg/datapath/linux/ipsec/cell.go]: This conflicts with the docs above. I also don't see where the "default to 50%" logic is.

File: pkg/datapath/linux/ipsec/cell.go (hunk 2 above)
Comment (adkafka) [inline pkg/datapath/linux/ipsec/cell.go]: Should we have a check that ensures the jitter < key rotation duration. If the jitter is longer, it will lead to errors IIUC.

File: pkg/datapath/linux/ipsec/cell_test.go [hunk 4]
```diff
@@ -328,3 +329,43 @@ func TestPrivileged_TestIPSecCell(t *testing.T) {
...
+			expectedJitter:   0,
+		},
+		{
+			name:             "explicit jitter is used",
+			rotationDuration: 10 * time.Second,
+			configuredJitter: 2 * time.Second,
+			expectedJitter:   2 * time.Second,
+		},
+		{
+			name:             "jitter can exceed roation duration if configured",
```
Comment (adkafka) [inline pkg/datapath/linux/ipsec/cell_test.go]: Is this a condition we want to allow for?

File: pkg/datapath/linux/ipsec/cell.go (hunk 3 above)
Comment (shavyshetty) [reply pkg/datapath/linux/ipsec/cell.go]: I initially started with implementing a default jitter of 50% of the rotation duration, but it would mean that jitter is always on, even for clusters that may not want it. I later switched to a default jitter of 0s, which effectively means disabled. I am now adding 50% as a suggested value.  I will update the comment to reflect this.

File: pkg/datapath/linux/ipsec/cell.go (hunk 2 above)
Comment (shavyshetty) [reply pkg/datapath/linux/ipsec/cell.go]: Good call, let me add tests as well.

Comment (shavyshetty) [reply pkg/datapath/linux/ipsec/cell.go]: Line 109 is accurate, but I can see where the confusion can come from, addressed those concerns here: https://github.com/cilium/cilium/pull/44263#discussion_r2788820794

**Comment by shavyshetty** [comment]
@Artyop, Can you please trigger the tests for me?

File: Documentation/security/network/encryption-ipsec.rst [hunk 5]
```diff
@@ -251,6 +251,14 @@ to all clusters in the mesh. You might need to increase the transition time to
 allow for the new keys to be deployed and applied across all clusters,
```
Comment (smagnani96) [inline Documentation/security/network/encryption-ipsec.rst:254]: Commits such as `Merge branch 'main' into feat/ipsec-key-rotation-jitter` must not be present.
We kindly ask authors to rebase their PRs on top of main, so that we avoid having these type of commits, and keep a clean git history.
Any change to address reviewer's feedback or adjust a feature added in one of the previous commits in the PR, should happen in the same commit for the same reason.

File: Documentation/security/network/encryption-ipsec.rst [hunk 6]
```diff
@@ -251,6 +251,14 @@ to all clusters in the mesh. You might need to increase the transition time to
...
 which you can do with the agent flag ``ipsec-key-rotation-duration``.
 
+In large clusters, key rotation can cause a "thundering herd" effect where all
+agents detect the key change simultaneously and update their CiliumNode resources
+at once, potentially overwhelming the Kubernetes API server. To mitigate this, 
+you can configure a random jitter delay before loading new keys using the Helm value
+``encryption.ipsec.keyRotationJitter`` or the agent flag ``ipsec-key-rotation-jitter``. 
+This is disabled by default. A recommended value is 50% of 
+``ipsec-key-rotation-duration`` (e.g., ``2m30s`` for the default 5m).
```
Comment (smagnani96) [inline Documentation/security/network/encryption-ipsec.rst]: I'm not particularly convinced by the jitter solution either, but I understand there could be problems.
Why do we want the jitter to be user-configurable rather than adding it automatically? (rand between 0 and keyrotation)

**Review by Artyop — CHANGES_REQUESTED** [review]
Thanks for the PR ! Please remove the merge commits and perform a rebase on main if you want to catch up the latests commits in your branch. Cilium doesn't do merge commits, they tend to pollute the tree for nothing :)

**Review by smagnani96 — CHANGES_REQUESTED** [review]

File: Documentation/security/network/encryption-ipsec.rst (hunk 6 above)
Comment (pchaigno) [reply Documentation/security/network/encryption-ipsec.rst]: I'd also prefer an automatic jitter. This is the sort of parameter that will be hard to configure because the appropriate value isn't obvious to most users.

Comment (pchaigno) [reply Documentation/security/network/encryption-ipsec.rst]: I wouldn't pick between 0 and maxRotation though. AFAIU, that would mean some nodes get a much smaller window (ex., if it picks something close to maxRotation). I'd prefer the random interval to be small compared to the key rotation duration.

I'm also not against increasing the default value of the max. key rotation if that makes things easier.

Comment (smagnani96) [reply Documentation/security/network/encryption-ipsec.rst]: We could go as suggested by the user with a value between `[0, maxRotation/2]` (or even `[0, min(1m, maxRotation/2)]`).

> I'm also not against increasing the default value of the max. key rotation if that makes things easier.

You mean it could help with the above computation? I'm not sure we should touch this 🤔

Comment (adkafka) [reply Documentation/security/network/encryption-ipsec.rst]: I agree, choosing `[0, maxRotation/2]` seems simpler than having an additional configuration dial to tune.

Comment (shavyshetty) [reply Documentation/security/network/encryption-ipsec.rst]: Based on the preferences in this thread, I updated the logic such that jitter will be between [0, maxRotation/2] if enabled. I kept the option as an opt-in for clusters rather than enabling jitter by default. Let me know if you all disagree with this.

**Review by joestringer — APPROVED** [review]
Docs changes look good to me, though I wonder why turn the feature off by default. I'll defer those details to @cilium/ipsec owners though.

nit: Please clean up the commit message.

Currently it shows like this:

```
ipsec: Add jitter to key file watcher to prevent thundering herd . Fixes#: 42721
Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>

fixed documentation indents

Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>

ipsec: Add key rotation jitter to prevent thundering herd. Fixes: https://github.com/cilium/cilium/issues/42721

Signed-off-by: Sharvil Shetty <shavyshetty@users.noreply.github.com>

ipsec: Add jitter to key file watcher to prevent thundering herd . Fixes#: 42721

Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>

ipsec: Add key rotation jitter to prevent thundering herd. Fixes: https://github.com/cilium/cilium/issues/42721

Signed-off-by: Sharvil Shetty <shavyshetty@users.noreply.github.com>

ipsec: Add jitter to key file watcher to prevent thundering herd . Fixes#: 42721

Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>

ipsec: Add key rotation jitter to prevent thundering herd. Fixes: https://github.com/cilium/cilium/issues/42721

Signed-off-by: Sharvil Shetty <shavyshetty@users.noreply.github.com>
```

It should be more like:
```
ipsec: Add jitter to key file watcher to prevent thundering herd 

We observed problems in our cluster where [...]

In order to fix this, add jitter into each agent to spread the
control plane load over a longer period so that the control plane
is not overwhelmed by many updates at the same time.

Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>
```

**Comment by shavyshetty** [comment]
> Docs changes look good to me, though I wonder why turn the feature off by default. I'll defer those details to @cilium/ipsec owners though.
> 
> nit: Please clean up the commit message.
> 
> Currently it shows like this:
> 
> ```
> ipsec: Add jitter to key file watcher to prevent thundering herd . Fixes#: 42721
> Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>
> 
> fixed documentation indents
> 
> Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>
> 
> ipsec: Add key rotation jitter to prevent thundering herd. Fixes: https://github.com/cilium/cilium/issues/42721
> 
> Signed-off-by: Sharvil Shetty <shavyshetty@users.noreply.github.com>
> 
> ipsec: Add jitter to key file watcher to prevent thundering herd . Fixes#: 42721
> 
> Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>
> 
> ipsec: Add key rotation jitter to prevent thundering herd. Fixes: https://github.com/cilium/cilium/issues/42721
> 
> Signed-off-by: Sharvil Shetty <shavyshetty@users.noreply.github.com>
> 
> ipsec: Add jitter to key file watcher to prevent thundering herd . Fixes#: 42721
> 
> Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>
> 
> ipsec: Add key rotation jitter to prevent thundering herd. Fixes: https://github.com/cilium/cilium/issues/42721
> 
> Signed-off-by: Sharvil Shetty <shavyshetty@users.noreply.github.com>
> ```
> 
> It should be more like:
> 
> ```
> ipsec: Add jitter to key file watcher to prevent thundering herd 
> 
> We observed problems in our cluster where [...]
> 
> In order to fix this, add jitter into each agent to spread the
> control plane load over a longer period so that the control plane
> is not overwhelmed by many updates at the same time.
> 
> Signed-off-by: Sharvil Shetty <sharvilshetty.200@gmail.com>
> ```

ACK on the commit messages, I thought I cleaned them up, but it looks like not. Will do. 

About the disabled by default, I left a note in the comment thread: 
> Based on the preferences in this thread, I updated the logic such that jitter will be between [0, maxRotation/2] if enabled. I kept the option as an opt-in for clusters rather than enabling jitter by default. Let me know if you all disagree with this. 
https://github.com/cilium/cilium/pull/44263#discussion_r2805957668

My thought is that smaller clusters won't encounter this issue, and cluster owners might prefer an immediate rotation instead. We only started seeing issues after reaching a certain size. I am flexible to go either way will wait for others to chime in.

**Comment by joestringer** [comment]
> My thought is that smaller clusters won't encounter this issue, and cluster owners might prefer an immediate rotation instead.

We do have some logic in `pkg/backoff` that scales time intervals depending on cluster size, might be another option to consider if the raw jitter numbers are too much. Though maybe that is a configuration not for the jitter but more for the `maxRotation` value(?)

Heads up a couple of the smoke tests fails, which will require you to update the PR. The failures list the commands to fix these things up at the end:
https://github.com/cilium/cilium/actions/runs/22002029026/job/63579007443?pr=44263
https://github.com/cilium/cilium/actions/runs/22002029056/job/63579029836?pr=44263

File: Documentation/security/network/encryption-ipsec.rst (hunk 6 above)
Comment (pchaigno) [reply Documentation/security/network/encryption-ipsec.rst]: If we go for `[0, maxRotation/2]`, then we've just halved the rotation duration for all existing users in the worst case. Not great. That's why I'd rather have the max. jitter small compared to the key rotation duration. So maybe `[0, maxRotation/10]`. If that doesn't work because it's too small, then I'd suggest to increase the rotation duration.

**Review by smagnani96 — CHANGES_REQUESTED** [review]
Follow-up of [the comment](https://github.com/cilium/cilium/pull/44263#discussion_r2792455792): I discussed this briefly with Paul. We believe we can always enable this behavior using a relatively small jitter delay, such as [0, keyRotationDuration/10].
I'd kindly ask to:

1. adjust code accordingly
2. remove the agent/helm flag to enable/disable: would not be easy and intuitive to use, let's always add a small jitter.
3. adjust Doc accordingly

**Comment by shavyshetty** [comment]
> Follow-up of [the comment](https://github.com/cilium/cilium/pull/44263#discussion_r2792455792): I discussed this briefly with Paul. We believe we can always enable this behavior using a relatively small jitter delay, such as [0, keyRotationDuration/10]. I'd kindly ask to:
> 
> 1. adjust code accordingly
> 2. remove the agent/helm flag to enable/disable: would not be easy and intuitive to use, let's always add a small jitter.
> 3. adjust Doc accordingly

Sorry, I missed @pchaigno's comment from a couple of days ago. I am not sure I understand this part: 
> If we go for [0, maxRotation/2], then we've just halved the rotation duration for all existing users in the worst case. Not great. 

Just curious, how does this effectively half the rotation duration? And if we instead do /10, would that mean we have reduced the rotation duration by a factor of 10? 

In the absence of jitter, the max rotation duration is respected, so how does adding a delay ( jitter) to keyfilewatcher alter that behavior and affect the validity of the old key?

**Comment by smagnani96** [comment]
@shavyshetty The worst case is when `jitter = maxRotation/2`, which is `2min30sec` in the default case with `maxRotation=5min`. This means nodes realize there's a new key only after 2min30sec. As a result, the rotation procedure has only `5min - 2min30sec` before we cleanup old state/policies referring to the previous key.

In large clusters this might not be enough, for this reason we're suggesting using a maximum jitter relatively small wrt the maximum key rotation time.

**Comment by adkafka** [comment]
> @shavyshetty The worst case is when `jitter = maxRotation/2`, which is `2min30sec` in the default case with `maxRotation=5min`. This means nodes realize there's a new key only after 2min30sec. As a result, the rotation procedure has only `5min - 2min30sec` before we cleanup old state/policies referring to the previous key.
> 
> In large clusters this might not be enough, for this reason we're suggesting using a maximum jitter relatively small wrt the maximum key rotation time.

This logic change sounds good to me.

I do want to note that there is some natural jitter already, because in kubernetes, an update to a secret / configmap is not instant across all nodes. It takes some time for all kubelet instances to notice the change, and make the update. In my experience, it spreads the changes out over ~90 seconds.

While it is a bit confusing to have jitter at two levels (one at kubelet, one introduced in this PR), I think it is still the right call. For large clusters, we likely want a larger jitter window & a larger maxRotation.

**Comment by shavyshetty** [comment]
Thanks for your reviews @Artyop, @adkafka, @joestringer, @pchaigno, and @smagnani96. I think I have now incorporated all your feedback, based on our collective understanding of the trade-offs for each option we explored. 
I appreciate your willingness to incorporate this change into Cillium. 

Please let me know if any further changes are needed for some CI tests.

**Comment by Artyop** [comment]
@shavyshetty you need to go fmt to have the go checks to pass

File: pkg/datapath/linux/ipsec/cell.go [hunk 7]
```diff
@@ -120,6 +120,15 @@ func (c Config) DNSProxyInsecureSkipTransparentModeCheckEnabled() bool {
 	return c.DNSProxyInsecureSkipTransparentModeCheck
 }
 
+// MaxKeyRotationJitter returns the maximum jitter duration to apply after
+// detecting a key file change before loading new keys. Jitter is set to 1/10
+// of the key rotation duration to prevent thundering herd on the K8s API server
```
Comment (smagnani96) [inline pkg/datapath/linux/ipsec/cell.go]:
```suggestion
// of the key rotation duration to prevent thundering herd on the K8s API server
```

File: pkg/datapath/linux/ipsec/cell.go [hunk 8]
```diff
@@ -120,6 +120,15 @@ func (c Config) DNSProxyInsecureSkipTransparentModeCheckEnabled() bool {
 	return c.DNSProxyInsecureSkipTransparentModeCheck
 }
 
+// MaxKeyRotationJitter returns the maximum jitter duration to apply after
+// detecting a key file change before loading new keys. Jitter is set to 1/10
+// of the key rotation duration to prevent thundering herd on the K8s API server 
+// while ensuring agents have sufficient time to load new keys before the old
```
Comment (smagnani96) [inline pkg/datapath/linux/ipsec/cell.go]:
```suggestion
// while ensuring agents have sufficient time to load new keys before the old
```

**Review by smagnani96 — CHANGES_REQUESTED** [review]
Two minor nits to please `golangci-lint`, then LGTM.
PR is way in a better shape than before, thanks for sticking with our feedback.

**Review by smagnani96 — APPROVED** [review]
LGTM, thanks.
@pchaigno if you want double-check last impl.

**Review by Artyop — APPROVED** [review]

**Review by adkafka — APPROVED** [review]

**Comment by shavyshetty** [comment]
@smagnani96 @pchaigno @joestringer 
Hey folks, just checking if anything is pending on my end for the PR to be merged?

**Review by smagnani96 — APPROVED** [review]
> @smagnani96 @pchaigno @joestringer
> Hey folks, just checking if anything is pending on my end for the PR to be merged?

LGTM, I'll start the tests 👍🏼

File: Documentation/security/network/encryption-ipsec.rst [hunk 9]
```diff
@@ -254,6 +254,14 @@ to all clusters in the mesh. You might need to increase the transition time to
 allow for the new keys to be deployed and applied across all clusters,
 which you can do with the agent flag ``ipsec-key-rotation-duration``.
 
+In large clusters, key rotation can cause a "thundering herd" effect where all
+agents detect the key change simultaneously and update their CiliumNode resources
+at once, potentially overwhelming the Kubernetes API server. To mitigate this, 
+Cilium applies a random jitter delay before loading new keys. The jitter is
+randomly selected from ``[0, keyRotationDuration/10]``, ensuring agents have
```
Comment (pchaigno) [inline Documentation/security/network/encryption-ipsec.rst:261]: The new interval also needs to be updated in the commit description.

Nit: You also have a `'` after the `Fixes:` tag.

File: Documentation/security/network/encryption-ipsec.rst
Comment (pchaigno) [inline Documentation/security/network/encryption-ipsec.rst:251]: (Commenting on the commit description.)

Could you please use a real email address in the Signed-off-by? These are required in case of license changes, for which we'll need to collect agreement from authors.

**Review by pchaigno — COMMENTED** [review]
The code looks good to me. Just a couple minor things to fix on the commit description.

File: Documentation/security/network/encryption-ipsec.rst
Comment (shavyshetty) [reply Documentation/security/network/encryption-ipsec.rst:251]: @pchaigno  Done!

**Comment by joestringer** [comment]
Thanks for the contribution @shavyshetty :tada:
